### PR TITLE
Refactor Differential Entropy for Mutual Information Classification

### PR DESCRIPTION
**Pull request by atavory**
== RELEASE NOTES ==

General Changes
* Move strategy factory code from udf to base class of strategy
* Add population counts/weights to reservoir samples

Together with #13203, this forms the basis for mutual information classification udfs #13163.

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/aggregate.rst [hunk 1]
```diff
@@ -572,7 +573,7 @@ That is, for a random variable :math:`x`, they approximate
 
 .. math ::
 
-    H(x) = - \int x \log_2\left(f(x)\right) dx,
+    h(x) = - \int x \log_2\left(f(x)\right) dx,
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/functions/aggregate.rst]: why? Also, put documentation changes into their own commit

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramJacknifeStateStrategy.java [hunk 2]
```diff
@@ -130,25 +137,19 @@ private static double getHoldOutEntropy(
...
-
     @Override
     public long getEstimatedSize()
     {
         return histogram.estimatedInMemorySize();
     }
 
     @Override
-    public int getRequiredBytesForSerialization()
+    public int getRequiredBytesForSpecificSerialization()
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramJacknifeStateStrategy.java]: The base class both writes the "code" identifying the subclass, as well as calls the subclass. Hence there is the required bytes for serialization with and without this constant. The `getRequiredBytesForSerialization` includes the size of the code, as well as the size needed by the specific subclass, which is returned by the abstract `getRequiredBytesForSpecificSerialization`.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java [hunk 3]
```diff
@@ -34,9 +34,10 @@ public void testIllegalMaxSamples()
     @Test
     public void testGetters()
     {
-        WeightedDoubleReservoirSample sample = new WeightedDoubleReservoirSample(200);
+        WeightedDoubleReservoirSample reservoir = new WeightedDoubleReservoirSample(200);
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java]: It seems more precise to call it a reservoir (of samples), than a (single) sample. Before this, there was code that seemed strange in the tests: there was a reservoir called a "sample", and then another data structure holding the expected samples, called "samples". It would stand to reason, therefore, that "sample" would hold a single something of which "samples" would hold several, but that was not the case. Currently, both "reservoir" and "samples" hold one or more "sample", which makes sense (to me).

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java [hunk 4]
```diff
@@ -25,35 +31,200 @@
...
+            DifferentialEntropyStateStrategy strategy,
+            DifferentialEntropyStateStrategy otherStrategy)
+    {
+        verify(strategy.getClass() == otherStrategy.getClass(),
+                format("In combine, %s != %s", strategy.getClass().getSimpleName(), otherStrategy.getClass().getSimpleName()));
+
+        strategy.mergeWith(otherStrategy);
+    }
+
+    DifferentialEntropyStateStrategy cloneEmpty();
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java]: It's used in #13163. I could move it there, but, as we discussed, this PR is basically changes that are needed for #13163, so not sure why to move this one in particular.

File: presto-docs/src/main/sphinx/functions/aggregate.rst (hunk 1 above)
Comment (atavory) [reply presto-docs/src/main/sphinx/functions/aggregate.rst]: #13163 is the first diff which combines functions of discrete entropy with those of differential entropy. Differential entropy is more precisely indicated by an $h$, not $H$, but that is sometimes ignored when on its own. When combining both discrete and differential entropy (as the docs of #13163 do or will do), it's important to differentiate between the two.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java [hunk 5]
```diff
@@ -25,35 +31,200 @@
...
+
+    double getTotalPopulationWeight();
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java]: It's used in #13163. I could move it there, but, as we discussed, this PR is basically changes that are needed for #13163, so not sure why to move this one in particular.

**Comment by rschlussel** [comment]
To clarify, I'm not suggesting moving changes to the other PR.  I'm asking if you can separate out the different logical changes in this PR into different commits so it's easier to review.  You can have multiple commits in a single PR (all the commits on a branch since master will be included in the PR), and github lets you review a PR commit by commit, so breaking this down into more granular changes would be very very helpful.

 Examples of independent logical changes for this pr would be: add mehod x to the interface + implementations, the changes in choosing a strategy (and can you provide an overview of what's going on there/ overall goal.  i don't quite understand it),  random variable name changes to improve clarity.

The easiest way to break apart your commit in git would probably be `git reset HEAD^` and then `git add -p` to select the changes you want for each commit.

**Comment by atavory** [comment]
> The easiest way to break apart your commit in git would probably be `git reset HEAD^` and then `git add -p` to select the changes you want for each commit.

Worked great, thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateFactory.java [hunk 6]
```diff
@@ -104,9 +104,6 @@ public DifferentialEntropyStateStrategy getStrategy()
         @Override
         public long getEstimatedSize()
         {
-            if (strategy == null) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateFactory.java]: is it no longer possible for strategy to be null?

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java [hunk 7]
```diff
@@ -58,6 +60,7 @@ public void testMany()
 
         long streamLength = 1_000_000;
         for (int i = 0; i < streamLength; ++i) {
+            assertEquals(reservoir.getTotalPopulationWeight(), (double) i);
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java]: nit: you shouldn't need to cast int to double.  Java should do it automatically.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyAggregation.java [hunk 8]
```diff
@@ -114,17 +77,12 @@ public static void input(
...
-                    format("Expected UnweightedReservoirSampleStateStrategy, got: %s", strategy.getClass().getSimpleName()));
-        }
-        strategy.validateParameters(size, sample);
-        strategy.add(sample, 1.0);
+        DifferentialEntropyStateStrategy strategy = DifferentialEntropyStateStrategy.getStrategy(
+                state.getStrategy(),
+                size,
+                sample);
+        state.setStrategy(strategy);
+        strategy.add(sample);
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyAggregation.java]: It is now done internally by the strategy base class. Since the point of this diff is to remove duplication between users of the strategy class, I'd like to remove the onus of passing the 1.0 from the callers of the class to the class itself internally.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/differentialentropy/AbstractTestReservoirAggregation.java [hunk 9]
```diff
@@ -32,12 +32,12 @@ protected String getFunctionName()
     @Override
     public Double getExpectedValue(int start, int length)
     {
-        assertTrue(length < MAX_SAMPLES);
+        assertTrue(2 * length < MAX_SAMPLES);
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/differentialentropy/AbstractTestReservoirAggregation.java]: Removed.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java (hunk 7 above)
Comment (atavory) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/reservoirsample/TestWeightedDoubleReservoirSample.java]: Just a point of interest: it passes with this cast, and fails without it. Bypassed it with a tolerance parameter (which is more correct here in any case).

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java [hunk 10]
```diff
@@ -79,7 +83,6 @@ private void addWithAdjustedWeight(double sample, double adjustedWeight)
     {
         if (count < samples.length) {
             samples[count] = sample;
-            weights[count] = adjustedWeight;
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java]: Great catch, thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramJacknifeStateStrategy.java [hunk 11]
```diff
@@ -170,4 +170,10 @@ public DifferentialEntropyStateStrategy clone()
     {
         return new FixedHistogramJacknifeStateStrategy(this);
     }
+
+    @Override
+    public DifferentialEntropyStateStrategy cloneEmpty()
+    {
+        return new FixedHistogramJacknifeStateStrategy(histogram.getBucketCount(), histogram.getMin(), histogram.getMax());
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramJacknifeStateStrategy.java]: Added them also to `FixedDoubleHistogram`.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java [hunk 12]
```diff
@@ -74,6 +75,13 @@ public void add(double sample, double weight)
         histogram.add(sample, weight);
     }
 
+    public double getTotalPopulationWeight()
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java]: Checked other instances in the code for this too.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java [hunk 13]
```diff
@@ -75,6 +75,7 @@ public void add(double sample, double weight)
         histogram.add(sample, weight);
     }
 
+    @Override
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java]: put this in the previous commit.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/differentialentropy/AbstractTestReservoirAggregation.java [hunk 14]
```diff
@@ -32,7 +32,7 @@ protected String getFunctionName()
     @Override
     public Double getExpectedValue(int start, int length)
     {
-        assertTrue(2 * length < MAX_SAMPLES);
+        assertTrue(length < MAX_SAMPLES);
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/differentialentropy/AbstractTestReservoirAggregation.java]: move this change to the previous commit

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java [hunk 15]
```diff
@@ -83,6 +83,7 @@ private void addWithAdjustedWeight(double sample, double adjustedWeight)
     {
         if (count < samples.length) {
             samples[count] = sample;
+            weights[count] = adjustedWeight;
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java]: put this in the previous commit

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java [hunk 16]
```diff
@@ -45,9 +45,9 @@
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedDoubleBreakdownHistogram.class).instanceSize();
 
-    private int bucketCount;
-    private double min;
-    private double max;
+    private final int bucketCount;
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java]: either put the `final` changes in their own commit, or rename this commit to indicate that it's general "clean up"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleHistogram.java [hunk 17]
```diff
@@ -37,9 +37,9 @@
 {
     public static class Bucket
     {
-        private final double left;
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleHistogram.java]: I don't think you meant to delete this.

**Comment by caithagoras** [comment]
@atavory @rschlussel Please enclose the release notes in a pair of triple backticks(```) in the future. Thanks!

**Comment by rschlussel** [comment]
Thanks @caithagoras.  These change are user facing, so actually no release notes needed.  Sorry for the trouble!
